### ISSUE TITLE
EntityLinker: linking from value of a custom extension 

### ISSUE THREAD

**Issue by kabirkhan** [explosion/spaCy#4981]
## Feature description

I've been thinking about Entity Linking on custom knowledge bases that don't have as many available aliases as Wikidata. There's a bit of a gap in the current implementation where the `ent.text` is assumed to be an alias in the `KnowledgeBase` and looked up directly. While this works fine for Wikidata it would be nice to allow the user to define a custom attribute they would like the linker to use instead of just the text.

As an example, if my NER model extracts an Entity with an Alias that is slightly different from that of any in my `KnowledgeBase`, the linking step won't work. Or I'll have to add a lot of variations of the alias to my `KnowledgeBase` to make the linking step work.

For example, the below code doesn't set the kb_id for "Machine learning" because the "l" in "learning" is not capitalized.

```python
nlp = spacy.load('custom_model')
print(nlp.get_pipe('entity_linker').kb.get_alias_strings())
["Machine Learning", "Artificial Intelligence", "AI"]
# 
text = "Machine learning is a subcategory of AI."
doc = nlp(text)

print([(e.text, e.label_, e.kb_id_) for e in doc.ents])
# [('Machine learning', 'SKILL', None), ('AI', 'SKILL', 'a1')]
```

This is a pretty obvious example but you could extend it to more generic and likely cases.

For example:

```python
print(kb.get_alias_strings())
# ["bioinformatics", "research"]
```

if the NER model extracts "researched" and "bioinformatic", they cannot be properly linked.

My proposal is allowing the user to define what attribute they want to use for looking up the alias in the `kb` using the `kb.get_candidates` method. This could be exposed as a pipeline config variable similar to how the EntityRuler accepts a `phrase_matcher_attr` variable.

This would allow the user to have a custom component before entity linking that can compute some string similarity function on the entity text to all the aliases in the `kb` to get all aliases above a threshold and then run the `kb.get_candidates` method for all of them.

So when we get candidates in the `predict` method we can check the alias attribute if the user set it in the config.

Something like:
```python
...
alias = getattr(ent, self.cfg.get(alias_attr, ''))
if not alias:
    alias = getattr(ent._, self.cfg.get(alias_attr, ''), ent.text))
candidates = self.kb.get_candidates(alias)
...
```

Happy to contribute this change but I thought I'd start a discussion first in case I'm overlooking something. (I think it would work fine to keep the training code the same but not 100% sure)
Really the above code is the core change that would be required. Actually setting the custom alias attr based on some similarity search between the `ent.text` and the aliases in the `kb` would live in a custom component.

## Could the feature be a [custom component](https://spacy.io/usage/processing-pipelines#custom-components) or [spaCy plugin](https://spacy.io/universe)?
If so, we will tag it as [`project idea`](https://github.com/explosion/spaCy/labels/project%20idea) so other users can take it on.

**Comment by svlandeg** [comment]
You're right that the current implementation deliberately cares about capitalization and small lexical variants because 1) we have a lot of that variants data from Wikipedia, and 2) they often really matter (think Apple vs. apple). I definitely see that we should be able to be more "fuzzy" for smaller corpora or corpora that aren't as strictly maintained as Wikipedia though.

Referring back to your original idea (https://github.com/explosion/spaCy/issues/4572#issuecomment-549563594), I think I like that proposal better that would allow `get_candidates` to be reimplemented / changed via a registered function or some other mechanism. That seems like a more general solution than what you're proposing here (match on a specific attribute). I'd love to hear your insights if you've looked into implementing either of these for your specific use-case(s), and ofcourse any help with a PR orso is also appreciated. 

We're currently working on spaCy v.3 following the recent `thinc` [release](https://thinc.ai/), and the idea is to have more flexible mechanisms (via config files) to change the inner workings of our models (NN architecture, hyperparameters, etc). It would definitely fit in that philosophy to be able to define your own `get_candidates` method too.

> This would allow the user to have a custom component before entity linking that can compute some string similarity function on the entity text to all the aliases in the kb to get all aliases above a threshold and then run the kb.get_candidates method for all of them.

This is for sure going to be pretty hard to implement efficiently. Right now the alias hash is used for a very fast lookup. As soon as you add string similarity computations, for **each** entity you want to resolve, that could potentially become a speed issue. But I haven't tested this yet.

Happy to keep brainstorming about this! Sorry for now following up in the earlier thread (Issue #4572) but I was offline for a month after that and haven't spent much time on the NEL module since I got back. We're still definitely keen to keep improving upon it though, and have it as a solid component in v.3.

**Comment by svlandeg** [comment]
> I think it would work fine to keep the training code the same but not 100% sure

Yes - the training does not care about aliases or synonyms of the entity, it only looks at the context of the word (e.g. surrounding sentence) and compares that to the entity definition (e.g. Wikipedia description) and gives a score. This score can then be augmented with the score from the `get_candidates` method that matches the mention with possible candidates.

**Comment by svlandeg** [comment]
I think this is now more generally addressed by PR #5876 so I'll close this issue (GH should have done that automatically but didn't?)
